### ISSUE TITLE
db: ingestion can break sequence number invariant

### ISSUE THREAD

**Issue by bananabrick** [cockroachdb/pebble#2196]
Ingestions have a [prepare](https://github.com/cockroachdb/pebble/blob/master/ingest.go#L753) step where we look for memtable overlap with the ingestion, and an [apply](https://github.com/cockroachdb/pebble/blob/master/ingest.go#L778) step where we move the ingested sstables into the lsm.

I have an inkling that detecting overlap like this in the prepare step isn't enough.

Consider the case where the lsm has a single memtable `M`. In the `prepare` step, the ingestion doesn't detect overlap with `M`. Once the `apply` step starts to run, the ingesting goroutine isn't holding any mutex(`db.mu` or `commitPipeline.mu`). Now, batches, which contain keys which overlap with the ingestion, with sequence numbers higher than the ingestion sequence number can get applied to `M`, which could force a flush on `M`. So, we could end up flushing `M`, before the apply step of the ingestion finishes. Let's say the sstables generated due to the flush are called `S`.

In this case, the ingested sstables could be layered on top of `S` in the apply step. But the ingested sstables have keys which overlap with `S`, and have sequence numbers lower than `S`. Thus inverting the sequence number invariant of the lsm.

Note that the `d.ingestApply` function within the `apply` function does grab `db.mu`, but there's a window where no mutex is held.

**Comment by sumeerbhola** [comment]
After re-reading the code it does look like a bug.
We do want flushes to proceed for the memtables we want flushed due to overlap. But we don't want the current mutable memtable after this loop https://github.com/cockroachdb/pebble/blob/6ca7956d84c1ce37036c015a4069aafc2f31e516/ingest.go#L763-L773 has finished, from being flushed (which will block future mutable memtables from also being flushed). So after that loop we could increment `memTable.writerRefs` and decrement it in `DB.ingestApply` after we grab `d.mu.versions.logLock()`. Once we have the lock on the `versionSet` we can allow the flush to proceed since it won't be able to apply the version edit until after this ingest.
 
But what prevents two concurrent ingestions from racing each other and allocating seqnums in one order Ingest1 before Ingest2, and then Ingest2 grabbing the `d.mu.versions.logLock()` before Ingest1? If they have overlapping files, we will place them such that Ingest1 files are at a higher level than Ingest2 but that violates the seqnum invariant. This should be

**Comment by jbowens** [comment]
> But what prevents two concurrent ingestions from racing each other and allocating seqnums in one order Ingest1 before Ingest2, and then Ingest2 grabbing the d.mu.versions.logLock() before Ingest1?

I think this is prevented by the spin loop in AllocateSeqNum: https://github.com/cockroachdb/pebble/blob/6ca7956d84c1ce37036c015a4069aafc2f31e516/commit.go#L319

The second ingest will wait for the first ingest’s sequence number to be published before continuing.

**Comment by sumeerbhola** [comment]
Good point about that spinning loop. That is going to spin for a while since it needs to wait for manifest IO. This is worse than waiting for memtable writes which don't wait for the WAL write in order to publish.

**Comment by jbowens** [comment]
This unit test exercises the bug:

```
diff --git a/ingest_test.go b/ingest_test.go
index 38031118..8d5a1ff3 100644
--- a/ingest_test.go
+++ b/ingest_test.go
@@ -9,11 +9,13 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"math"
 	"os"
 	"sort"
 	"strconv"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -23,10 +25,12 @@ import (
 	"github.com/cockroachdb/pebble/internal/base"
 	"github.com/cockroachdb/pebble/internal/errorfs"
 	"github.com/cockroachdb/pebble/internal/keyspan"
+	"github.com/cockroachdb/pebble/internal/manifest"
 	"github.com/cockroachdb/pebble/internal/rangekey"
 	"github.com/cockroachdb/pebble/internal/testkeys"
 	"github.com/cockroachdb/pebble/objstorage"
 	"github.com/cockroachdb/pebble/objstorage/objstorageprovider"
+	"github.com/cockroachdb/pebble/record"
 	"github.com/cockroachdb/pebble/sstable"
 	"github.com/cockroachdb/pebble/vfs"
 	"github.com/kr/pretty"
@@ -1424,6 +1428,130 @@ func TestIngestMemtablePendingOverlap(t *testing.T) {
 	require.NoError(t, d.Close())
 }
 
+// TestIngestMemtableOverlapRace is a regression test for the race described in
+// #2196. If an ingest that checks for overlap with the mutable memtable and
+// finds no overlap, it must not allow overlapping keys with later sequence
+// numbers to be applied to the memtable and the memtable to be flushed before
+// the ingest completes.
+//
+// This test operates by committing the same key concurrently:
+//   - 1 goroutine repeatedly ingests the same sstable writing the key `foo`
+//   - n goroutines repeatedly apply batches writing the key `foo` and trigger
+//     flushes.
+//
+// After a while, the database is closed and the manifest is verified. Version
+// edits should contain new files with monotonically increasing sequence
+// numbers, since every flush and every ingest conflicts with one another.
+func TestIngestMemtableOverlapRace(t *testing.T) {
+	mem := vfs.NewMem()
+	el := MakeLoggingEventListener(DefaultLogger)
+	d, err := Open("", &Options{
+		FS: mem,
+		// Disable automatic compactions to keep the manifest clean; only
+		// flushes and ingests.
+		DisableAutomaticCompactions: true,
+		// Disable the WAL to speed up batch commits.
+		DisableWAL:    true,
+		EventListener: &el,
+		// We're endlessly appending to L0 without clearing it, so set a maximal
+		// stop writes threshold.
+		L0StopWritesThreshold: math.MaxInt,
+		// Accumulating more than 1 immutable memtable doesn't help us exercise
+		// the bug, since the committed keys need to be flushed promptly.
+		MemTableStopWritesThreshold: 2,
+	})
+	require.NoError(t, err)
+
+	// Prepare a sstable `ext` setting foo=v.
+	f, err := mem.Create("ext")
+	require.NoError(t, err)
+	w := sstable.NewWriter(objstorageprovider.NewFileWritable(f), sstable.WriterOptions{})
+	for _, k := range []string{"foo"} {
+		require.NoError(t, w.Set([]byte(k), []byte("v")))
+	}
+	require.NoError(t, w.Close())
+
+	var done atomic.Bool
+	const numSetters = 2
+	var wg sync.WaitGroup
+	wg.Add(numSetters + 1)
+
+	untilDone := func(fn func()) {
+		defer wg.Done()
+		for !done.Load() {
+			fn()
+		}
+	}
+
+	// Ingest in the background.
+	totalIngests := 0
+	go untilDone(func() {
+		filename := fmt.Sprintf("ext%d", totalIngests)
+		require.NoError(t, mem.Link("ext", filename))
+		require.NoError(t, d.Ingest([]string{filename}))
+		totalIngests++
+	})
+
+	// Apply batches and trigger flushes in the background.
+	wo := &WriteOptions{Sync: false}
+	var localCommits [numSetters]int
+	for i := 0; i < numSetters; i++ {
+		i := i
+		v := []byte(fmt.Sprintf("v%d", i+1))
+		go untilDone(func() {
+			// Commit a batch setting foo=vN.
+			b := d.NewBatch()
+			require.NoError(t, b.Set([]byte("foo"), v, nil))
+			require.NoError(t, b.Commit(wo))
+			localCommits[i]++
+			d.AsyncFlush()
+		})
+	}
+	time.Sleep(100 * time.Millisecond)
+	done.Store(true)
+	wg.Wait()
+
+	var totalCommits int
+	for i := 0; i < numSetters; i++ {
+		totalCommits += localCommits[i]
+	}
+	m := d.Metrics()
+	tot := m.Total()
+	t.Logf("Committed %d batches.", totalCommits)
+	t.Logf("Flushed %d times.", m.Flush.Count)
+	t.Logf("Ingested %d sstables.", tot.TablesIngested)
+	require.NoError(t, d.CheckLevels(nil))
+	require.NoError(t, d.Close())
+
+	// Replay the manifest. Every flush and ingest is a separate version edit.
+	// Since they all write the same key and compacitons are disabled, sequence
+	// numbers of new files should be monotonically increasing.
+	dbDesc, err := Peek("", mem)
+	require.NoError(t, err)
+	require.True(t, dbDesc.Exists)
+	f, err = mem.Open(dbDesc.ManifestFilename)
+	require.NoError(t, err)
+	defer f.Close()
+	rr := record.NewReader(f, 0 /* logNum */)
+	var largest *fileMetadata
+	for {
+		r, err := rr.Next()
+		if err == io.EOF || err == record.ErrInvalidChunk {
+			break
+		}
+		require.NoError(t, err)
+		var ve manifest.VersionEdit
+		require.NoError(t, ve.Decode(r))
+		for _, f := range ve.NewFiles {
+			require.Equal(t, 0, f.Level)
+			if largest != nil && largest.LargestSeqNum > f.Meta.LargestSeqNum {
+				t.Fatalf("previous largest file %s has sequence number > next file %s", largest, f.Meta)
+			}
+			largest = f.Meta
+		}
+	}
+}
+
 type ingestCrashFS struct {
 	vfs.FS
 }
```

**Comment by sumeerbhola** [comment]
> The second ingest will wait for the first ingest’s sequence number to be published before continuing.

Working on the proof sketch this morning, I came back to this. I think this is insufficient. Even if both ingests land in L0, there is nothing preventing the later seqnum ingest's files to be compacted out of L0 before the earlier ingest completes, so we'll have inverted the levels. And more commonly, the later seqnum ingest will land in a level lower than L0 and then in ingestApply we place the file of the earlier seqnum ingest at a higher level, again inverting the levels.

**Comment by jbowens** [comment]
> Working on the proof sketch this morning, I came back to this. I think this is insufficient. 

I might be misunderstanding, but I think it's sufficient because this waiting on the sequence number publishing happens before _prepare_. I think it completely synchronizes `Ingests` with one another, so that one that acquired an earlier sequence number must have already successfully applied (linked into the LSM, or now linked into the flushable queue) before the ingest with a later sequence number checks for overlap with the flushable queue.

**Comment by sumeerbhola** [comment]
> I think it completely synchronizes Ingests with one another

You are right

**Comment by sumeerbhola** [comment]
I suspect we have a different race bug that breaks the seqnum-stack from the proof sketch.

Key k#t1 in S1 and k#t2 in S2, where t1 < t2. S1 is added to L0 before S2. S1 is a flush and S2 is an ingest, and the latter is also added to L0. If LargestSeqNum(S1) > LargestSeqNum(S2), it will invert the sub-level ordering in L0 after S2 is added.

Consider the sequence of memtables M1, M2, Mn, where Mn is the mutable memtable, when the ingest request of S2 arrives. Consider k#t1 is in Mn (the following example also works for k#t1 in Mi, as long as Mn also becomes full and flushes together with Mi). S2 waits for the flush of Mn. I think Mn can concurrently get a write, which will be assigned seqnum > t2 (say a key different than k), that is added to it before the flush.

Sequence of steps for S2:
1. Acquire commitPipeline.mu
2. Allocate S2 seqnum (t2)
3. Acquire DB.mu
4. Check for memtable overlap and compute Mn as the youngest memtable that overlaps
5. Schedule flush
6. Release DB.mu
7. Release commitPipeline.mu
8. Another write gets added to Mn with seqnum > t2
9. Wait for Mn flush to finish
10. Acquire DB.mu
11. versionSet.logLock
12. Ingest at appropriate level
13. versionSet.logUnlock
14. Release DB.mu

**Comment by sumeerbhola** [comment]
I think I am wrong about this being a problem for Mn containing k#t1, since we do
```
			if mem.flushable == d.mu.mem.mutable {
				err = d.makeRoomForWrite(nil)
			}
```
But if the conflict with a i < n, this can happen, since Mn can flush with that Mi. I think the solution is the same as https://github.com/cockroachdb/pebble/issues/2196#issuecomment-1359561774 -- we want a barrier preventing the mutable memtable from flushing, until after the ingest has completed.

**Comment by jbowens** [comment]
Fixed in #2486 and backports #2500 (23.1) and #2501 (22.2).
